### ISSUE TITLE
Unable to find fundamental matrix for regular markov chains

### ISSUE THREAD

**Issue by povinsahu1909** [sympy/sympy#20189]
Currently, it is not possible to calculate the fundamental matrix for regular markov chains as of now ``` fundamental_matrix()``` was only implemented for the absorbing Markov chains. 
For example
```python
>>> T = Matrix([[0.5, 0.25, 0.25], [0.5, 0, 0.5], [0.25, 0.25, 0.5]])
>>> Y = DiscreteMarkovChain("Y", [0, 1, 2], T)
>>> Y.is_regular()
True
>>> Y.fundamental_matrix()
DiscreteMarkovChain(Y, FiniteSet(0, 1, 2), Matrix([
[ 0.5, 0.25, 0.25],
[ 0.5,    0,  0.5],
[0.25, 0.25,  0.5]]))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-5-bd1859c21b86> in <module>
----> 1 Y.fundamental_matrix()

~/project/sympy(main)/sympy/sympy/stats/stochastic_process_types.py in fundamental_matrix(self)
    638         I = eye(Q.shape[0])
    639         if (I - Q).det() == 0:
--> 640             raise ValueError("Fundamental matrix doesn't exists.")
    641         return ImmutableMatrix((I - Q).inv().tolist())
    642 

ValueError: Fundamental matrix doesn't exists.
```
But correct answer should be
``` 
Matrix([
[  1.14666666666667, 0.04, -0.186666666666667],
[              0.08, 0.84, 0.0799999999999999],
[-0.186666666666667, 0.04,   1.14666666666667]])
```

**Comment by czgdp1807** [comment]
Can you please cite the source for this example, like in which document/web page did you come across this example?

**Comment by povinsahu1909** [comment]
> Can you please cite the source for this example, like in which document/web page did you come across this example?

Page no. 54 of https://www.dartmouth.edu/~chance/teaching_aids/books_articles/probability_book/Chapter11.pdf

**Comment by Maelstrom6** [comment]
According to that source, the definition of the fundamental matrix is:

> Definition 11.3 For an absorbing Markov chain P, the matrix N = (I − Q)−1 is
called the fundamental matrix for P. The entry nij of N gives the expected number
of times that the process is in the transient state sj if it is started in the transient
state si.

They seem to secretly relax the "transient" parts of the definition later on for some reason.

and from Wikipedia
> A basic property about an absorbing Markov chain is the expected number of visits to a transient state j starting from a transient state i (before being absorbed).

So IMHO, that example should have a 0x0 fundamental matrix rather than a 3x3 fundamental matrix since every state there is recurrent. We have no transient states in that chain. The expected number of visits to each of those recurrent states is infinity.

**Comment by Maelstrom6** [comment]
Maybe this matrix should have a different name since it has a different definition compared to the typical fundamental matrix. They also seem to use Z rather than N after providing a different definition of it.

**Comment by povinsahu1909** [comment]
>They seem to secretly relax the "transient" parts of the definition later on for some reason.

I think both the quoted definition is same as fundamental matrix can be interpreted as the expected number of times the chain is in state j if it started in state i.

**Comment by povinsahu1909** [comment]
> They also seem to use Z rather than N after providing a different definition of it.

For this, I think that just to avoid confusion authors used two different symbols to denote the fundamental matrix i.e N for absorbing chains and Z for the regular markov chains.

**Comment by Maelstrom6** [comment]
> I think both the quoted definition is same as fundamental matrix can be interpreted as the expected number of times the chain is in state j if it started in state i.

Yes, they are the same. But I found later on that they change it to

> As in the regular case, we will call the inverse of the matrix I − P + W the
fundamental matrix for the ergodic chain with transition matrix P, and we will use
> Z to denote this fundamental matrix.

You are using this definition which is different from the original 2 definitions. For your example and the typical definition, the fundamental matrix should be a 0x0 matrix. For your example and the ergodic definition, the matrix is as you state. It is for this reason that I have added my suggestions in my second comment on the PR.
